SPI Engine offload support and ad7944/ad7380 mainline pre-review

dlech opened the pull request:
This is my work in progress so far on mainlining SPI Engine offload support. I have also created a new AD7944 driver for testing this. Most of this is still a work in progress, so I'm not looking for detailed review of everything quite yet.

The first 12 patches (everything before "dt-bindings: spi: adi,axi-spi-engine: add offload bindings") are just cleanups and improvements of the existing upstream SPI Engine driver. These patches should be ready to submit upstream ASAP as a series if there aren't any objections.

There are some additional non-offload related SPI Engine improvements starting with "spi: axi-spi-engine: populate xfer->effective_speed_hz", but those need more testing and cleanup before they will be ready for pre-review here and submission to the mailing lists. I plan to send these as a second series once the first has been finalized and picked up.

The offload support will be a 3rd series and will probably need to also include the ad7944 driver to show how consumers of the offload work. FYI, I have already run the offload devicetree bindings by Rob Herring on IRC, so I don't expect any major changes needed there. There are a few things to sort out with the HDL first (i.e. https://github.com/adi-ses/sea-misc/pull/56) before the details are finalized. 

But, the big picture is...
* Compared to the offload implementation in the ADI tree, a bit of the SPI Engine offload support is moved to the core SPI code as discussed in https://github.com/analogdevicesinc/linux/issues/2300#issuecomment-1774867382 which simplifies things a bit.
* I wrote the AD7944 driver so that it works with any SPI controller (without any offload mechanism). The reasoning being that this would ensure that the IIO ABI is "correct" and is not being bent to match offload requirements (and who knows, maybe someday someone would want to use this with a regular SPI controller with cyclic DMA transfers). 
* I implemented the PWM and DMA that are connected to the SPI Engine offload as a sort of standalone device that is both an IIO trigger and an IIO buffer (currently, the poorly titled "iio: trigger: add PWM trigger " patch).

You will also find a bunch of `//` comments in the later patches. This is intentional to remind myself of things that need more work before submitting upstream. If anyone has any bright ideas for addressing any of those TODOs though, I will gladly take suggestions since I haven't thought of good solutions to a lot of them yet.

nunojsa commented:
> The first 12 patches (everything before "dt-bindings: spi: adi,axi-spi-engine: add offload bindings") are just cleanups and improvements of the existing upstream SPI Engine driver. These patches should be ready to submit upstream ASAP as a series if there aren't any objections.

Alright, for now I will concentrate on those. I hope to have some time today or tomorrow so you can send them upstream ASAP. If there's no feedback on me till Thursday, please ping me.

threexc reviewed (commented):
Really minor comment - in the commit message for https://github.com/analogdevicesinc/linux/pull/2341/commits/8d057a84a482fc8b3a22b825893aeb0dbb15c400, you say "However, it did set any..." when it should be "However, it didn't set any..."

dlech commented:
> If there's no feedback on me till Thursday, please ping me.

ping @nunojsa

nunojsa commented:
> ping @nunojsa

Oh yeah, will do it tomorrow morning... Sorry for the delay.

nunojsa commented on Documentation/devicetree/bindings/spi/adi,axi-spi-engine.yaml:
```diff
@@ -0,0 +1,66 @@
+      #size-cells = <0>;
+
+      /* SPI devices */
+    };
```
nit: please better indent the example (I already got complains about it). IIRC, 4 spaces is the typical (but maybe 2 is also acceptable)

nunojsa commented on drivers/spi/spi-axi-spi-engine.c:
```diff
@@ -531,8 +531,6 @@ static void spi_engine_remove(struct platform_device *pdev)
 	struct spi_controller *host = platform_get_drvdata(pdev);
 	struct spi_engine *spi_engine = spi_controller_get_devdata(host);
 
-	spi_unregister_controller(host);
```
Since you're doing all this work, an extra patch would be to put the cleanup in `.remove()` in `devm_add_action_or_reset()` and get completely rid of the callback.

In fact, this could even be problematic because now you changed the order things get removed/cleaned which makes me wonder if the switch for devm should not be one single patch... Anyways, the things is, the .remove() hook will be called first and only then the devres_cleanup is done (on driver<->device unbind) [1] & [2]. So, now you're doing those writes while still having the controller registered...

[1]: https://elixir.bootlin.com/linux/latest/source/drivers/base/dd.c#L1272
[2]: https://elixir.bootlin.com/linux/latest/source/drivers/base/dd.c#L1278

nunojsa commented on drivers/spi/spi-axi-spi-engine.c:
```diff
@@ -517,6 +517,9 @@ static int spi_engine_probe(struct platform_device *pdev)
 	host->num_chipselect = 8;
 
+	if (host->max_speed_hz == 0)
+		return dev_err_probe(&pdev->dev, -EINVAL, "spi_clk rate is 0");
```
just wondering (did not checked)... is the rest of the probe() path using `dev_err_probe()`?

dlech replied:
No, the rest of probe just returns errors with no logging.

nunojsa commented on drivers/spi/spi-axi-spi-engine.c:
```diff
@@ -417,29 +441,46 @@ static int spi_engine_transfer_one_message(struct spi_controller *host,
+	ret = ida_alloc_range(&spi_engine->sync_ida, 0, U8_MAX, GFP_KERNEL);
+	if (ret < 0) {
+		kfree(p);
+		kfree(st);
```
maybe take a look at the cleanup.h for the FREE stuff to see if is something you can use here...

dlech replied:
In this particular instance, using `__free(kfree)`  would require significant changes to the rest of the function, so probably best the way it is.

nunojsa commented on drivers/spi/spi-axi-spi-engine.c:
```diff
@@ -417,29 +441,46 @@ static int spi_engine_transfer_one_message(struct spi_controller *host,
 	size_t size;
+	int ret;
+
+	st = kzalloc(sizeof(*st), GFP_KERNEL);
```
POne thing that worries me is that now we have one extra allocation. Did you noticed anything in terms of performance? (I know that for real performance we should use the offload engine :)). Still, just wondering...

dlech replied:
I will test it.

nunojsa commented on drivers/spi/spi-axi-spi-engine.c:
```diff
@@ -236,6 +245,13 @@ static int spi_engine_compile_message(struct spi_engine *spi_engine,
+			bits_per_word = xfer->bits_per_word;
+			spi_engine_program_add_cmd(p, dry,
+				SPI_ENGINE_CMD_WRITE(SPI_ENGINE_CMD_REG_XFER_BITS,
+					bits_per_word));
```
I'm wondering one thing... Is this a per core setting or will apply at a transfer level? What I'm wondering is that if we can really support transfers with different `bits_per_word` in one message. Maybe I'm just confused now...

dlech replied:
When not using the offload, it should be fine. Since this is creating a SPI Engine instruction, it is per-transfer.

I need to dig deeper to see what the implications are when using the offload, but practically, I would not expect anyone to be trying to stream different sized data at the same time from a single chip using the offload.

nunojsa commented on drivers/spi/spi-axi-spi-engine.c:
```diff
@@ -233,20 +236,28 @@ static int spi_engine_compile_message(struct spi_engine *spi_engine,
+				keep_cs = true;
+			} else {
+				if (!xfer->cs_off)
+					spi_engine_gen_cs(p, dry, spi, false);
```
a bit puzzled with the logic... if `cs_off` is not set shouldn't we assert the cs? Could you also explain the reasoning about checking the next messages? Sorry, I'm being lazy and not trying much to understand it....

dlech replied:
As mentioned in the commit message, this logic is copied from the generic `spi_transfer_one_message()` in `spi.c`, so it seems safe to assume it is the "correct"/desired behavior and has been well tested.

nunojsa reviewed (commented):
LGTM... Nice series. The one more worrying thing is the devm conversion. Other than that, just some things that likely I'm failing to understand.

So, go ahead :)

nunojsa commented:
I would also make a comment about the maintainers you set (me and Michael). I'm ok with that but the original developer was Lars, so maybe it would be a good idea to ask him if he's ok with and if he wants to get included in the list.

Also remembered one thing.... There's no MAINTAINERS entry for this driver AFAICT, so one patch for it would be nice (like the first one in the series :))

dlech commented:
Status update: the first round of patches has been sent upstream: https://lore.kernel.org/linux-spi/20231117-axi-spi-engine-series-1-v1-0-cc59db999b87@baylibre.com/

I will make another comment when the next round of patches after that is ready to look at again.

dlech commented:
I have updated this with the latest work in progress to facilitate the discussion in https://github.com/adi-ses/sea-misc/issues/55.

Also rebased on current spi/for-6.8 where some of these patches have already been applied.

dlech commented:
The second round of patches is ready for review (ping @nunojsa).

Since the last review, many of these patches are new or have significantly changed. Once we've had a few more eyes on these, I will send this as a series on the mailing lists.

098e80260ea5 spi: axi-spi-engine: fix CPHA
7bf66345de35 spi: axi-spi-engine: change return of spi_engine_compile_message() to void
8965c13b9787 spi: axi-spi-engine: populate xfer->effective_speed_hz
34cd0a362541 spi: axi-spi-engine: remove spi_engine_get_clk_div()
cb1d761ba2c1 spi: axi-spi-engine: fix sleep ticks calculation
a1cc53b78379 spi: axi-spi-engine: remove xfer arg from spi_engine_gen_sleep()
d874af4bd802 spi: axi-spi-engine: implement xfer->cs_change_delay
dfcf742ebc7c spi: axi-spi-engine: restore clkdiv at end of message
85d27515762f spi: axi-spi-engine: remove delay from CS assertion
27b9c56cbd60 spi: axi-spi-engine: add watchdog timer

nunojsa commented on drivers/spi/spi-axi-spi-engine.c:
```diff
@@ -132,7 +132,8 @@ static unsigned int spi_engine_get_config(struct spi_device *spi)
 		config |= SPI_ENGINE_CONFIG_CPOL;
-	if (spi->mode & SPI_CPHA)
+	/* SPI Engine definition of CPHA is unconventional. */
+	if (!(spi->mode & SPI_CPHA))
```
Is this code upstream? I guess so... Don't forget the fixes tag

dlech replied:
In our weekly call to today, it was decided that this should be considered an HDL bug and so I will drop this patch.

nunojsa reviewed (approved):
LGTM... nice series.

dlech commented on drivers/spi/spi-axi-spi-engine.c:
```diff
@@ -219,7 +219,7 @@ static void spi_engine_gen_cs(struct spi_engine_program *p, bool dry,
 }
 
-static int spi_engine_compile_message(struct spi_engine *spi_engine,
+static void spi_engine_compile_message(struct spi_engine *spi_engine,
```
In one of the patches, we add a `spi_engine_precompile_message()` function which would be the better place to check for error if needed. But for now, `__spi_validate()` does the error checking for us and I haven't found anything it misses yet.

threexc commented on drivers/spi/spi-axi-spi-engine.c:
```diff
@@ -260,12 +246,12 @@ static void spi_engine_compile_message(struct spi_engine *spi_engine,
 			spi_engine_program_add_cmd(p, dry,
 				SPI_ENGINE_CMD_WRITE(SPI_ENGINE_CMD_REG_CLK_DIV,
-					clk_div));
+					clk_div - 1));
```
Same idea here as with my other comment - is it worth being explicit (with a comment) about why subtracting is needed? It's already in the commit message but being more obvious may help for future work.

dlech replied:
Registers requiring the divider minus one seems commonplace to me, so I didn't see a need for a comment about this. Depends on the experience of the person reading it, I guess.

dlech commented on drivers/spi/spi-axi-spi-engine.c:
```diff
@@ -132,7 +132,8 @@ static unsigned int spi_engine_get_config(struct spi_device *spi)
 	if (spi->mode & SPI_CPOL)
 		config |= SPI_ENGINE_CONFIG_CPOL;
-	if (spi->mode & SPI_CPHA)
+	/* SPI Engine definition of CPHA is unconventional. */
```
Yeah, I was hoping to get confirmation from someone at ADI that this was either an intentional design choice or if it was an oversight, confirm that it should be fixed here rather than in HDL.

dlech replied:
we are dropping this patch and fixing the HDL instead.

commodo commented on drivers/iio/trigger/axi-spi-engine-offload-pwm-trigger.c:
```diff
@@ -0,0 +1,275 @@
+
+static const struct of_device_id axi_offload_pwm_trigger_match_table[] = {
+	{ .compatible = "adi,axi-spi-engine-offload-pwm-trigger-dma-output" },
+	{ },
```
nitpick: null terminators don't need a trailing comma

commodo commented on drivers/spi/spi-axi-spi-engine.c:
```diff
@@ -183,66 +204,101 @@ static void spi_engine_gen_xfer(struct spi_engine_program *p, bool dry,
 		mask ^= BIT(spi_get_chipselect(spi, 0));
 
-	spi_engine_program_add_cmd(p, dry, SPI_ENGINE_CMD_ASSERT(1, mask));
+	gpio_state->value = (mask & ~gpio_state->mask) | (gpio_state->value & gpio_state->mask);
```
maybe some bitops helpers could be used here?
`set_bit` or the like?

commodo commented on drivers/spi/spi-axi-spi-engine.c:
```diff
@@ -602,6 +620,225 @@ static int spi_engine_transfer_one_message(struct spi_controller *host,
+	if (reg >= SPI_ENGINE_MAX_NUM_OFFLOADS)
+		return ERR_PTR(-EINVAL);
+
+	struct spi_offload *offload = &spi_engine->offloads[reg];
```
does kernel guidelines allow C99 style variable declarations like this?

(it may have changed, just asking)

dlech replied:
It think they are technically allowed if you have a good reason, but still generally discourage. No reason for this one here, so good catch.

dlech commented on drivers/spi/spi-axi-spi-engine.c:
```diff
@@ -124,6 +158,8 @@ static void spi_engine_program_add_cmd(struct spi_engine_program *p,
 		p->instructions[p->length] = cmd;
 	p->length++;
+
+	// printk_ratelimited(KERN_DEBUG "%s: %04x\n", dry ? "D" : "W", cmd);
```
Yes, there is lots of this still. I've been trying to polish up a smaller series of patches at a time and then make a comment like https://github.com/analogdevicesinc/linux/pull/2341#issuecomment-1832849627 when they are ready for closer review before submitting upstream.

dlech commented on drivers/pwm/Makefile:
```diff
@@ -2,6 +2,7 @@
 obj-$(CONFIG_PWM)		+= core.o
 obj-$(CONFIG_PWM_SYSFS)		+= sysfs.o
 obj-$(CONFIG_PWM_AB8500)	+= pwm-ab8500.o
+obj-$(CONFIG_PWM_AXI_PWMGEN)	+= pwm-aix-pwmgen.o
```
FYI, there is a separate PR that improves the pwmgen driver at https://github.com/analogdevicesinc/linux/pull/2363

Looks like we got the name right there.

commodo commented:
> Since getting AD7944 working is depending on HDL changes to the SPI Engine, we've changed the strategy a bit and we are now going to use AD7380 as the first mainline driver to get offload support upstream. So to prepare for this, I have cleaned up the [ad738x.c](https://github.com/analogdevicesinc/linux/blob/0f65cf46403b978d101c9c7d1f842e05a3b9e008/drivers/iio/adc/ad738x.c#L102) driver from the ADI tree.
> 

yeah, that's a good strategy;
with all these FPGA elements (related to these ADCs/DACs) it's always easier to split them up a bit;
whenever sending a big-enough patchset, 1 comment will block a 10-20-patch-set;

dlech commented on drivers/iio/adc/ad7380.c:
```diff
@@ -0,0 +1,462 @@
+	 */
+	struct {
+		u16 raw[2];
+		s64 ts;
```
```suggestion
		s64 ts __aligned(8);
```

https://github.com/analogdevicesinc/linux/commit/89bf90c75fbc99275b49fcf5864a5e90b44f689c#r134463165

dlech commented on drivers/iio/adc/ad7380.c:
```diff
@@ -0,0 +1,462 @@
+			     chan->scan_type.realbits - 1);
+
+	return IIO_VAL_INT;
+
```
```suggestion
```

https://github.com/analogdevicesinc/linux/commit/89bf90c75fbc99275b49fcf5864a5e90b44f689c#r134463326

dlech commented on drivers/iio/adc/ad7380.c:
```diff
@@ -0,0 +1,462 @@
+		return ret;
+
+	/* SPI 1-wire mode */
+	ret = regmap_update_bits(st->regmap, AD7380_REG_ADDR_CONFIG2,
+				 AD7380_CONFIG2_SDO,
+				 FIELD_PREP(AD7380_CONFIG2_SDO, 1));
+	if (ret < 0)
+		return ret;
+
+	return 0;
```
```suggestion
	return regmap_update_bits(st->regmap, AD7380_REG_ADDR_CONFIG2,
				 AD7380_CONFIG2_SDO,
				 FIELD_PREP(AD7380_CONFIG2_SDO, 1));
```

https://github.com/analogdevicesinc/linux/commit/89bf90c75fbc99275b49fcf5864a5e90b44f689c#r134463511

dlech commented on drivers/iio/adc/ad7380.c:
```diff
@@ -0,0 +1,462 @@
+	if (ret)
+		return ret;
+
+	ret = ad7380_init(st);
```
```suggestion
	ret = ad7380_init(st);
	if (ret)
	       return ret;
```

https://github.com/analogdevicesinc/linux/commit/89bf90c75fbc99275b49fcf5864a5e90b44f689c#r134464368

commodo commented on drivers/iio/adc/ad7380.c:
```diff
@@ -0,0 +1,462 @@
+	st->chip_info = spi_get_device_match_data(spi);
+
+	st->vref = devm_regulator_get_optional(&spi->dev, "refio");
+	if (IS_ERR(st->vref))
```
it's not uncommon to print these messages about using internal/external regulator;

i have no preference on the matter;
some argue it's log spam;

nunojsa commented:
@dlech can we close this one?

dlech commented:
Yes, let's close.
